NIP-23 Revisions (relay list) (#164)
nostr-protocol/nips: 23.md

`draft` `optional` `author:fiatjaf` `author:cameri` `author:monlovesmango` `author:giszmo`

A special event with kind `10001`, meaning "relay list" is defined as a stringified content field containing an object where each member represents a relay the author uses.

This kind is similar to kind 2 in format, but would live at kind 10001 to discourage relay bloat and encourage more consistent implementation. 

The key for each object member is the `relay URL` and **must** omit the trailing slash. 

The tags array is empty.

The `read` value operates on the values of the [NIP-01](01.md) **filter** object, while the `write` rule operates on the values of the **event** object. 

Both `read` and `write` values are required to be set to either `true` or `false`. 

Both the `read` and `write` values are strings, to enable extensibility and type conflicts. 

When the `read` value equals the string `true`  then read from that relay, otherwise don't. 

When the `write` value equals the string `true`  then write to that relay, otherwise don't.

_Valid event.content_
```
"{\"wss://relayer.somewhere.xyz\":{\"read\": \"true\",\"write\": \"true\"}}"
```

> _Invalid event.content: trailing slash_
> ```
> "{\"wss://relayer.somewhere.xyz/\":{\"read\": \"true\",\"write\": \"true\"}}"
> ```

Reviewer: ```suggestion
```
I would drop this altogether. Let's not restrict which relay URLs are valid or not. If you can connect it is valid, if not, it isn't.

Author: The problem with that, is that then you will have duplicate relays in the list if one client adds trailing slashes, and another client does not.

Reviewer: I think you should make your app tolerant to these by treating the trailing slash and non slash urls the same instead of forcing this validation. I expect clients to send even false relays that you can't ever connect to because the ip or host is not public.

Author: Yes, absolutely, regardless of the standard things will happen sometimes. But by not explicitly stating trailing slash requirements, it will happen a lot more, and there will be no means to resolve. With a canonical NIP, resolution is inherent. 

Given the scenario

```
{
  "wss://somerelay/" : { "read": "true", write: "false" }
  "wss://somerela": { "read": "false" }
}
```

Now not only does a dev need to dedupe, but they also now need to merge these two arrays, and decide which one to listen to.

Given a strict standard, the compliant client would just remove the invalid record, because they know due to a canonical standard which is "valid".

Reviewer: We could normalize URLs using https://developer.mozilla.org/en-US/docs/Web/API/URL
```
var a = new URL('wss://relay')
var b = new URL('wss://relay/')

a.toString() === b.toString() // true
```

Author: @Cameri Let's wait for some others to weigh in.

Author: > We could normalize URLs using https://developer.mozilla.org/en-US/docs/Web/API/URL

How does this resolve the problem with **subjective validity?**

Author: Out of curiousity, what are the core concerns of the standard imposing trailing slashes? I can only think of one, which is that:

1. A server config may be configured with a backslash, and sanitizing the backslash could present issues (or vice versa if it were strict in the other direction) 

I'm sure you have a valid reason to be against it, but to me it is unclear.

Reviewer: I think it is not necessary to mandate slashes (or not) but I don't have a strong desire to block this NIP because of it. Please disregard my comments and proceed with your requirement.

Author: Would be good to get some more input from others. I'm going to reopen and we can revisit tomorrow. I wouldn't get started on my implementation until next week anyways.

Author: @mikedilger @staab Any thoughts?

Reviewer: Should we have the same restriction everywhere a relay is mentioned or just this particular NIP?

Reviewer 2: I do not think we should validate relay urls, any string is "valid", it's the client's responsibility to deduplicate, validate, catch errors, etc.

Author: @staab To be clear it's not validation, it would be part of the standard to avoid collisions. Have you considered the scenario I presented above and the implications on the user experience?

Reviewer 2: Nostr is just going to have all kinds of weird subjective validity problems due to its open nature, mandating a trailing slash is only whacking one of a million moles. I've had to be many times more defensive in my coding with Coracle than on any other project I've worked on. Permissiveness isn't easy, but the simplicity is worth the pain of implementation.

In your merging example, a client does indeed have to just pick a strategy, because there's no truly correct way to dedupe/merge. But I also think any method, including dropping both, picking first/last, merging forward, and merging backward are acceptable approaches because they approximate the intent, which is all any client can do in a distributed system. But I also think that for the most part it's just not going to be a problem because a) no one would do that on purpose, so a spec prohibiting it would be redundant with common sense, and b) users have access to their relay list and can manually de-dupe or correct things. Clients could also notify the user that something's wacky with their relays, which would solve the subjective validity issue, as users are the only authority on what relay settings should be.

Author: > Should we have the same restriction everywhere a relay is mentioned or just this particular NIP? 

@Cameri I'm not sure tbh, didn't think this one would be contentious, but it clearly is. At the end of the day my motivations are to get something through by consensus, even if I personally disagree with it. 

The rationale behind the relay format is holistic. Eliminate code bloat for devs and prevent poor UX with a clear out-route. 

I do see that it might not be necessary in the purest sense of the word, but it will probably result in a much cleaner implementation.

Reviewer 2: IOW, you're asking clients to be smart enough to contradict a user when they enter a trailing slash, rather than be smart enough to notice problems and as the user for help.

Author: > But I also think any method, including dropping both, picking first/last
Object keys can change position

> merging forward
This is a replaceable kind, so if I understood you correctly, merging forward isn't really an option here (unless by merging forward you mean overwriting user data with your client data) 

>merging backward are acceptable approaches because they approximate the inten which is all any client can do in a distributed system. 
I absolutely agree, and clarifying in a standard which format is "valid" and defensive programming are not mutually exclusive. 

> as users are the only authority on what relay settings should be.
Yes, but 99% of them will enter the relay through a tool/client, which can sanitize in a one liner.

Reviewer 2: By merging I was referring to https://github.com/nostr-protocol/nips/pull/164#discussion_r1068294779

With specs, I think they should be minimal or exhaustive. Mandating no trailing slash is neither, and nostr's ethos is one of minimalism.

Author: > By merging I was referring to https://github.com/nostr-protocol/nips/pull/164#discussion_r1068294779

There's no way to merge that non-destructively in a way that is logical. If there is guidance on a trailing slash, then dropping one of them in this scenario, is objectively justified. Maybe it starts a trailing vs no-trailing slash war, maybe the standard is amended to remove the guidance. 

> With specs, I think they should be minimal or exhaustive. Mandating no trailing slash is neither, and nostr's ethos is one of minimalism

Again, this isn't a mandate, it's technically a guidance. Virtually every other NIP has the language **must**, it doesn't detract from it's minimalism.

Author: > IOW, you're asking clients to be smart enough to contradict a user when they enter a trailing slash, rather than be smart enough to notice problems and as the user for help.

I'm actually asking the opposite, to handle the above scenario without guidance requires the client to be incredibly smart. To handle the scenario where the is guidance on the correct format is simply a rejection (1-2 lines of code). The client would also technically protecting the user from potential data loss

The only case that's been outlined, which I think @Cameri and you are touching on, is that enforcing trailing vs no trailing can result an edge case where a server is configured to only accept a trailing slash. 

And it is a completely valid case.

Author: Would be really good to get some more feedback from others. At this juncture, based on feedback it looks to me like this conversation is leaning towards removing relay URL formatting from the standard. Also, thanks for your input @staab!

Reviewer: Your last sentence on your first reply to my initial suggestion was that you wouldn't die on this hill. Are you suuuuure?
🤪
On a serious note: are relays with obvious path like /ws/ also follow this rule? Keeping in mind that /ws and /ws/ are distinct paths. Is the restriction only when the path is just /?

Relays can't reject these events because they don't inspect the content field. This is a client-only rule.

Author: Part of the process is arguing any points, hashing things out and clarifying any misconceptions. My last comment reads 

> At this juncture, based on feedback it is clear we are leaning towards removing relay URL formatting from the spec.

Unless new information comes up, I probably don't have anything left to say on the topic of a relay URL format

> On a serious note: are relays with obvious path like /ws/ also follow this rule? Keeping in mind that /ws and /ws/ are distinct paths. Is the restriction only when the path is just /?

If this rule somehow gets through, which at this point it does not appear it will, it would apply to any url. 

> Relays can't reject these events because they don't inspect the content field. This is a client-only rule.

Yes, this was expressed above.

Author: I was going to leave this unresolved until tomorrow, if there's no further feedback, I'll happily merge in your changes :D

_Invalid event.content: missing `write` member_
```
"{\"wss://relayer.somewhere.xyz\":{\"read\": \"true\"}}"
```

> ### Purposes

Reviewer: ```suggestion
### Use Cases
```

This NIP serves two purposes: (1) backup and interoperability of relay lists and relay specific rules between clients; and (2) sharing of relay URLs between users.

The first use case is meant to make it so users can open their client -- or different clients -- in different devices and have their list of relays automatically fetched from a default relay and start using their relays list without having to set everything up again.

> For the second purpose, if any client decides to, they can show to the user what relays other users are using, suggest that or automatically add these to the user's relay list. The possibility of sharing a list of relays in standardized format is good for spreading information about relays and contributes to the censorship-resistance of the network.

Reviewer: ```suggestion
Another use case is to display which relays other users are using, suggest connecting to them or automatically add these to the user's relay list. The possibility of sharing a list of relays in standardized format is good for spreading information about relays and contributes to the censorship-resistance of the network.
```
